browser/src/dashboard/Dashboard.tsx

```diff
@@ -93,102 +110,269 @@ export class Dashboard extends React.Component<DashboardProps, DashboardState> {
+                    connection: {
+                      ...this.state.connection,
+                      isFetching: false,
+                      status: ConnectionStatus.DISCONNECTED
+                    }
+                  })
+                })
+            } else if (
+              !this.state.connection.lastFetchedEventsTimestamp || 
+              this.state.connection.lastFetchedEventsTimestamp && (now - this.state.connection.lastFetchedEventsTimestamp) > this.props.tickInterval * 1000
+            ) {
+              this.setState({
+                connection: {
+                  ...this.state.connection,
+                  isFetching: true
+                }
+              })
+              const from = moment(this.state.connection.lastWindowTimestamp).utc()
+              const to = moment(windowNow).add(this.props.tickInterval * 3, 'seconds').utc()
+              const differenceDrift = from.valueOf() - windowNow
+              const windowSize = to.valueOf() - from.valueOf()
+
+              // also have a kill switch on the window size
+              if (differenceDrift < this.props.tickInterval * 1000) {
+                log(now, windowNow, `Detected too much drift between the window and current time window size(${windowSize}ms) drift(${differenceDrift}ms)`)
+                this.setState({
+                  connection: {
+                    ...this.state.connection,
+                    isFetching: false,
+                    status: ConnectionStatus.DISCONNECTED
+                  },
+                  lastConnectionTick: now
+                })
+                return
+              }
+              log(now, windowNow, `Fetch event window ${from.format(format)} ${to.format(format)} window size(${windowSize}ms) drift(${differenceDrift}ms)`)
+              fetchEventTicker(from, to)
+                .then((result) => {
+                  this.setState({
+                    queuedEvents: [
+                      ...this.state.queuedEvents,
+                      ...result.events
+                    ],
+                    connection: {
+                      ...this.state.connection,
+                      isFetching: false,
+                      lastFetchedEventsTimestamp: now,
+                      lastWindowTimestamp: to.valueOf()
+                    }
+                  })
+                })
+                .catch((error) => {
+                  log(now, windowNow, 'Error during fetching event window')
+                  this.setState({
+                    connection: {
+                      ...this.state.connection,
+                      isFetching: false,
+                      status: ConnectionStatus.DISCONNECTED
+                    }
+                  })
+                })
+            }
+            break
+          case ConnectionStatus.FAILED:
+            // unhook this tick listener, we don't need to continue evaluating
+            break
+          default:
+            break
+        }
```

Comment:
There is quite a lot going on in this one function, would be good to break it down a little in future so it doesn't explode someones head looking at it.

Reply:
Yeah agreed, it would be ideal if none of this happened in the top level dashboard and was instead managed by smaller scoped classes. I'm planning to move this out of this repo in a bit and i'll include all of these suggestions in that move